src/primitives/compact_action.rs

```diff
@@ -0,0 +1,181 @@
+//! Defines actions for Orchard shielded outputs and compact action for light clients.
+
+// Review hint: this file is largely derived from src/note_encryption.rs
+
+use core::fmt;
+
+use zcash_note_encryption::{note_bytes::NoteBytes, EphemeralKeyBytes, ShieldedOutput};
+
+use crate::{
+    action::Action,
+    note::{ExtractedNoteCommitment, Nullifier, Rho},
+};
+
+use super::{orchard_domain::OrchardDomain, orchard_primitives::OrchardPrimitives};
+
+impl<A, P: OrchardPrimitives> ShieldedOutput<OrchardDomain<P>> for Action<A, P> {
+    fn ephemeral_key(&self) -> EphemeralKeyBytes {
+        EphemeralKeyBytes(self.encrypted_note().epk_bytes)
+    }
+
+    fn cmstar(&self) -> &ExtractedNoteCommitment {
+        self.cmx()
+    }
+
+    fn cmstar_bytes(&self) -> [u8; 32] {
+        self.cmx().to_bytes()
+    }
+
+    fn enc_ciphertext(&self) -> Option<&P::NoteCiphertextBytes> {
+        Some(&self.encrypted_note().enc_ciphertext)
+    }
+
+    fn enc_ciphertext_compact(&self) -> P::CompactNoteCiphertextBytes {
+        P::CompactNoteCiphertextBytes::from_slice(
+            &self.encrypted_note().enc_ciphertext.as_ref()[..P::COMPACT_NOTE_SIZE],
+        )
+        .unwrap()
+    }
+}
+
+/// A compact Action for light clients.
+#[derive(Clone)]
+pub struct CompactAction<P: OrchardPrimitives> {
+    nullifier: Nullifier,
+    cmx: ExtractedNoteCommitment,
+    ephemeral_key: EphemeralKeyBytes,
+    enc_ciphertext: P::CompactNoteCiphertextBytes,
+}
+
+impl<P: OrchardPrimitives> fmt::Debug for CompactAction<P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "CompactAction")
+    }
+}
+
+impl<A, P: OrchardPrimitives> From<&Action<A, P>> for CompactAction<P>
+where
+    Action<A, P>: ShieldedOutput<OrchardDomain<P>>,
+{
+    fn from(action: &Action<A, P>) -> Self {
+        CompactAction {
+            nullifier: *action.nullifier(),
+            cmx: *action.cmx(),
+            ephemeral_key: action.ephemeral_key(),
+            enc_ciphertext: action.enc_ciphertext_compact(),
+        }
+    }
+}
+
+impl<P: OrchardPrimitives> ShieldedOutput<OrchardDomain<P>> for CompactAction<P> {
+    fn ephemeral_key(&self) -> EphemeralKeyBytes {
+        EphemeralKeyBytes(self.ephemeral_key.0)
+    }
+
+    fn cmstar(&self) -> &ExtractedNoteCommitment {
+        &self.cmx
+    }
+
+    fn cmstar_bytes(&self) -> [u8; 32] {
+        self.cmx.to_bytes()
+    }
+
+    fn enc_ciphertext(&self) -> Option<&P::NoteCiphertextBytes> {
+        None
+    }
+
+    fn enc_ciphertext_compact(&self) -> P::CompactNoteCiphertextBytes {
+        P::CompactNoteCiphertextBytes::from_slice(self.enc_ciphertext.as_ref()).unwrap()
```

Comment:
_From str4d ([r2599562825](https://github.com/zcash/orchard/pull/471#discussion_r2599562825))_

```suggestion
        self.enc_ciphertext.clone()
```
and alter the `OrchardPrimitives::CompactNoteCiphertextBytes` associated type bounds as necessary to be `Clone`.

Reply:
FIXED